### ISSUE TITLE
Minor issue with table clip rect

### ISSUE THREAD

**Issue by BobbyAnguelov** [ocornut/imgui#6765]
Really minor thing but for tables that have vertical scrolling the table contents are drawn over the borders.

![image](https://github.com/ocornut/imgui/assets/10767490/f83b488b-3891-4fc2-9017-2840138381ca)

If you run the examples and go to the tables section in the demo and look at the "vertical scrolling, with clipping" example, you can see the issue.

Zoomed in for more clarity: 
![image](https://github.com/ocornut/imgui/assets/10767490/fd08a76b-4a62-4405-8dd5-60b58b2ca2e5)

**Comment by v-ein** [comment]
Try to replace `ImGuiTableFlags_BordersOuter` with `ImGuiTableFlags_BordersOuterH` and see if the issue persists. I bet it will only happen when you have **both** horizontal and vertical outer borders. That's because the table is optimized to render such borders as a rectangle, but the rectangle gets wrong coordinates :).

**Comment by v-ein** [comment]
BTW the issue might become more of a nuisance when you place opaque widgets into the table - they get painted over the borders, making the right-side and bottom borders piecewise... depending on the contents, might look pretty ugly.

**Comment by v-ein** [comment]
So... to make it clear - here's how the table renders its outer borders:

```cpp
if ((table->Flags & ImGuiTableFlags_BordersOuter) == ImGuiTableFlags_BordersOuter)
{
    inner_drawlist->AddRect(outer_border.Min, outer_border.Max, outer_col, 0.0f, 0, border_size);
}
else if (table->Flags & ImGuiTableFlags_BordersOuterV)
{
    inner_drawlist->AddLine(outer_border.Min, ImVec2(outer_border.Min.x, outer_border.Max.y), outer_col, border_size);
    inner_drawlist->AddLine(ImVec2(outer_border.Max.x, outer_border.Min.y), outer_border.Max, outer_col, border_size);
}
else if (table->Flags & ImGuiTableFlags_BordersOuterH)
{
    inner_drawlist->AddLine(outer_border.Min, ImVec2(outer_border.Max.x, outer_border.Min.y), outer_col, border_size);
    inner_drawlist->AddLine(ImVec2(outer_border.Min.x, outer_border.Max.y), outer_border.Max, outer_col, border_size);
}
```

You see, when both borders are enabled, the border is rendered as a rectangle with bottom-right corner in `outer_border.Max` - only that the bottom-right corner is exclusive (contrary to top-left which is inclusive), and therefore `outer_border.Max` actually lies **outside** of the rectangle frame. When only horizontal or only vertical borders are enabled, the line is drawn through `outer_border.Max`, making it inclusive. This causes a 1-pixel offset for right and bottom sides, depending on which flags are used.

This seems to render the borders properly:

```cpp
inner_drawlist->AddRect(outer_border.Min, outer_border.Max + ImVec2(1, 1), outer_col, 0.0f, 0, border_size);
```

Note: I've only tested this on Windows and not sure how it behaves on other backens, including high-DPI screens.

**Comment by ocornut** [comment]
Hello,

I investigated this and your fix @v-ein seems right, pushed it as a340718. Thank you!

Apologies I didn't investigate this and #3752 earlier: some of the border code in tables is subtly tricky because it closely relates to cliprect matching and ImDrawCmd merging constraints. However making this specific changes passes my regression tests in ImGuiTestSuite and didn't alter ImDrawCmd count.

While investigating this I also spotted another long-time bug there in some situations the top-most outer border would be drawn twice, once with TableBorderLight (incorrectly), once with TableBorderStrong (correctly). The earlier would show underneath when TableBorderStrong has alpha under 1.0f. This is fixed with 5a483c2.

One remaining subtle issue that CellBg would be displayed underneath the left-most and top-most border, again only visible when TableBorderStrong.a < 1.0f.

**Comment by BobbyAnguelov** [comment]
![image](https://github.com/ocornut/imgui/assets/10767490/4c80465b-44a2-4091-b446-00d620e1c7bf)

Hi, so I was able to sync to latest and this issue only seems to be fixed for the bottom border, the top border still has the issue as you can see in the screenshot taken from the demo app.

**Comment by BobbyAnguelov** [comment]
Would the remaining issues account for this?

![image](https://github.com/ocornut/imgui/assets/10767490/2490c11e-902f-4cd4-bf83-073417610bc6)

It basically still draws the cell contents on top of the border in the example app. 

It's not so bad in the demo app, but it my case (with my theme colors), it gives the impression a bunch of my UI is kinda broken: 

![image](https://github.com/ocornut/imgui/assets/10767490/b1645a74-fe94-45c6-b279-6e2c5f8cbef6)

**Comment by v-ein** [comment]
I believe it's a different issue. Horizontal borders are drawn within the cells, with top border of each row being potentially overlaid by cell contents. In most cases cell padding prevents contents from overlaying the border. However, if you set vertical padding to zero, or like in your case, scroll the table, there's a good chance that the border gets painted over.

In contrast to that, vertical borders are rendered outside the cells, i.e. the layout mechanism does account for space taken by borders between columns and by vertical sides of the outer border. Therefore they are not overlaid by contents.

Here's a picture I recently made explaining the difference between how vertical and horizontal borders are rendered:

![imgui-borders](https://github.com/ocornut/imgui/assets/104593001/8ae48db5-f56c-46b0-a93e-2eefe195a222)

**Comment by ocornut** [comment]
The previous fix caused an issue #7428

I've reverted it (revert: a131c3e61) and I'm proposing another fix (864a2bf), you can see me here toggling the change on/off (_after_ the revert which is ):

![table_clip_6765](https://github.com/user-attachments/assets/ffe35f37-b959-45ec-b941-3c10c63a3ff7)

Toggling the change on/off before the revert (so you can see the line moving to)

![table_clip_6765_before_revert](https://github.com/user-attachments/assets/33b74466-0e75-4ccc-80ec-87d552ce5e2f)

It's unfortunately a little bit complicated for many reasons so I'm hoping I didn't break anything else, but I want a tested with that toggle in many areas and it seems generally betters, and it passes my tests for draw-call matching/merging.
Hope I did get it right this time!
